### DISCUSSION TITLE
Improve ergonomics on `identity::Keypair`

### DISCUSSION THREAD

**Discussion by drHuangMHT** [libp2p/rust-libp2p#3623] [Ideas]
In #3350 we've made the decision to make `Keypair` opaque. But I think there's more to be done. Currently deprecating keypair variants makes it harder to know what's inside. You have to try those methods one by one to see what it is. This won't be a problem if you're not importing and exporting keys. Instead of trying out many different methods to get the key imported or exported(which is annoying), we can unify them into two with one responsible for importing(bytes to `Keypair`) and one for exporting(`Keypair` to bytes). Matching can be done internally instead of letting users handle this. Plus, it's not a piece of performance critical code.

**Comment by thomaseizinger** [comment]
We have `to_protobuf_encoding` which we could extend to also export other key types.

Two questions:

- Why do you need other keys than ed25519?
- Do you need the type for something else or just exporting?

**Reply by drHuangMHT** [reply]
>     Why do you need other keys than ed25519?

I would say I'm providing support for them simply because they are here in the library. The upstream supports these keys then why not me? It doesn't hurt.
 
>     Do you need the type for something else or just exporting?

I think there's nothing else to do with the keys other than import and export. They are practically the same to the application.

**Reply by drHuangMHT** [reply]
Protobuf encoding makes exchanging keys between libp2p applications easy, but I'm thinking of other cases where they don't necessarily have support for protobuf encoding.

**Reply by thomaseizinger** [reply]
> Protobuf encoding makes exchanging keys between libp2p applications easy, but I'm thinking of other cases where they don't necessarily have support for protobuf encoding.

The format is meant to be opaque. As long as you can import what was exported, isn't that good enough?

**Reply by drHuangMHT** [reply]
I would rather not have a ton of keys to manage on my keychain simply because the cover is different. At least we can provide a way to get the binary representation of the key.

**Reply by thomaseizinger** [reply]
> I would rather not have a ton of keys to manage on my keychain simply because the cover is different.

What do you mean by that?

> At least we can provide a way to get the binary representation of the key.

The only one that has a standardized binary representation is ed25519 and we encode to that already.

From the perspective of our library, it makes sense to provide an encode and a decode function that are compatible to allow for persisting keys between node restarts.

We reuse existing standards where possible (see ed25519). For the others where there isn't an obvious standard, I'd say the format is to be treated as opaque.

Is your requirement that you want to import those keys into other applications?

**Reply by drHuangMHT** [reply]
> What do you mean by that?

Which one do you prefer, one key per identity or one key per application? The former one would greatly improve interoperability while the latter creates little islands. I want things to be interconnected.

> Is your requirement that you want to import those keys into other applications?

No. Not at all. It's not a requirement. Just as I said before, if `libp2p` supports those types of key, I'll make sure those keys are also supported by me, that's it. I'm writing an application and I'll try my best to provide support for what the library supports. If the implementation is somewhat broken, I would rather not have it than broken. If it's worthy of doing, do it in the best way.

**Reply by thomaseizinger** [reply]
> > What do you mean by that?
> 
> Which one do you prefer, one key per identity or one key per application? The former one would greatly improve interoperability while the latter creates little islands. I want things to be interconnected.

In previous projects I was involved in, we had a master seed and deterministically derived the network identity from that using a HKDF.

I am not sure if this is what you mean by "one key per application" but we've found that system to work really well.

It allowed us to store the master seed in whatever format and generate secrets like wallet seeds or network identities without storing numerous keys that the user has to backup.

> > Is your requirement that you want to import those keys into other applications?
> 
> No. Not at all. It's not a requirement. Just as I said before, if `libp2p` supports those types of key, I'll make sure those keys are also supported by me, that's it. I'm writing an application and I'll try my best to provide support for what the library supports. If the implementation is somewhat broken, I would rather not have it than broken. If it's worthy of doing, do it in the best way.

We support interacting with all key types, i.e. decoding their public key. If you can, I'd limit the application to only use ed25519 keys.

I can see your point though, perhaps we should remove support for the other private key types. @mxinden thoughts?

**Reply by drHuangMHT** [reply]
> > > What do you mean by that?
> > 
> > 
> > Which one do you prefer, one key per identity or one key per application? The former one would greatly improve interoperability while the latter creates little islands. I want things to be interconnected.
> 
> In previous projects I was involved in, we had a master seed and deterministically derived the network identity from that using a HKDF.

I don't see the point of doing that though. May you explain why?

> I am not sure if this is what you mean by "one key per application" but we've found that system to work really well.

I don't think it is. I would like to put myself in my users' shoes and try to guess what they would do. I'll try to ensure good user experience. My application is targeting the mass and I'll make sure it's free, not only the app itself, but also what the users can do.

I'm terribly sorry if my words don't make sense. I'm not a native speaker. I'm Chinese. Maybe we see things a bit differently and I am kind of a weirdo myself.

**Reply by drHuangMHT** [reply]
> > At least we can provide a way to get the binary representation of the key.
> 
> The only one that has a standardized binary representation is ed25519 and we encode to that already.

If other keys are not even standardized, then why should we ship them? That's really confusing. I think we should remove them now and add them back after they are standardized.

**Reply by thomaseizinger** [reply]
> > > > What do you mean by that?
> > > 
> > > 
> > > Which one do you prefer, one key per identity or one key per application? The former one would greatly improve interoperability while the latter creates little islands. I want things to be interconnected.
> > 
> > 
> > In previous projects I was involved in, we had a master seed and deterministically derived the network identity from that using a HKDF.
> 
> I don't see the point of doing that though. May you explain why?

HKDF stands for hierarchically deterministic key derivation function. The point is, if you derive your libp2p keys deterministically, you never need to export them, you just generate them again on start-up.

> > I am not sure if this is what you mean by "one key per application" but we've found that system to work really well.
> 
> I don't think it is. I would like to put myself in my users' shoes and try to guess what they would do. I'll try to ensure good user experience. My application is targeting the mass and I'll make sure it's free, not only the app itself, but also what the users can do.

I have little context about what your application does but from I gather, you want to allow your users to configure what key type they want to use? Personally, I'd just make a decision for yours users. When in doubt, less config options are better IMO.

Any modern libp2p application (across languages, i.e. `go-libp2p` etc) can understand ED25519 keys, meaning there is no benefit for your users to choose one of these other key types.

**Reply by thomaseizinger** [reply]
> > > At least we can provide a way to get the binary representation of the key.
> > 
> > 
> > The only one that has a standardized binary representation is ed25519 and we encode to that already.
> 
> If other keys are not even standardized, then why should we ship them? That's really confusing. I think we should remove them now and add them back after they are standardized.

Standardized in the sense that there is a standard _outside_ of libp2p for how to encode those keys. We reuse that encoding, see https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#ed25519.

What I meant by non-standardized is that the libp2p team just chose a representation for the other keys, see https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#key-types. They are standardized across the libp2p ecosystem but not outside of that.

Simply exposing the "binary representation" of a key doesn't make much sense as that would be leaking internal details too much. Different languages or implementations may choose to represent those keys differently in-memory. That is why we have the specs repository to document an **export format** (see links above). That format is implement in the `{to,from}_protobuf_encoding` functions. Support for the other key types can be added, I opened an issue for this: https://github.com/libp2p/rust-libp2p/issues/3630

**Reply by drHuangMHT** [reply]
> When in doubt, less config options are better IMO.

Type `Keypair` has already abstracted some logic to eliminate the difference between those keys, they can be used as long as they're included in the variants. But we didn't further abstract the logic after we made the type opaque causing troubles when importing and exporting keys.   
Please let me be honest, I'm losing patients now because I have been trying to convey my point for so long. It happened before but this time I really get upset. Maybe I should get the code running before opening a discussion about it.

**Reply by drHuangMHT** [reply]
Think think think how would other apps save their keys. If there are standards for those kinds of keys, they *should* point out what the key *itself* looks like. I believe the actual key does not vary with the application.

**Reply by drHuangMHT** [reply]
Let me summarize with an analogy. I'm working in a distribution center where I should put packages into different ports according to tags on both. Then the distribution center announced that they are trying to prevent packages from being put into wrong ports. They removed the tags on the ports leaving me to guess where the package should go to. I proposed that a sorting machine should be built between me and those ports, given that both the packages and ports have tags that can be matched by machines. Then the manager questioned why I'm using so many ports. Why? Because they are here, the distribution center claimed that they can handle those packages so they should be prepared. That's what is happening when importing keys.  
I know not why you keep trying to drive the car off the lane. The two questions at the beginning are NOT related to implementation at all yet we still spend so much time digging into them.

**Reply by drHuangMHT** [reply]
I'll remain inactive for quite some time from now on because yesterday I blew up my physical server at dorm that is purchased using my parents' hard-earned money, the schoolwork is also coming at me. If I don't stop, I'll blow my self up, literally. I really don't know why I'm in such a situation.

**Reply by thomaseizinger** [reply]
> Type `Keypair` has already abstracted some logic to eliminate the difference between those keys, they can be used as long as they're included in the variants. But we didn't further abstract the logic after we made the type opaque causing troubles when importing and exporting keys.

Once https://github.com/libp2p/rust-libp2p/issues/3630 is implemented, any type of key can be serialized and thus exported and imported again. Is that sufficient for your usecase?

**Comment by drHuangMHT** [comment]
Also a pending draft PR that's probably worth noticing #2671.
